qiskit/qpy/binary_io/circuits.py

```diff
@@ -612,13 +720,22 @@ def write_circuit(file_obj, circuit, metadata_serializer=None):
     index_map["c"] = {bit: index for index, bit in enumerate(circuit.clbits)}
     for instruction in circuit.data:
         _write_instruction(instruction_buffer, instruction, custom_instructions, index_map)
-    file_obj.write(struct.pack(formats.CUSTOM_CIRCUIT_DEF_HEADER_PACK, len(custom_instructions)))
 
-    for name, instruction in custom_instructions.items():
-        _write_custom_instruction(file_obj, name, instruction)
+    with io.BytesIO() as custom_instructions_buffer:
+        new_custom_instructions = list(custom_instructions.keys())
+        while new_custom_instructions:
+            for name in new_custom_instructions:
```

Comment:
I have small concern here because `new_custom_instructions` value is updated by the loop itself (and it's not copied). This doesn't induce any edge case?

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/8055/commits/6c7279c5ac2ac0a969503c41254735c362c2d973